libraries/AP_Math/scurves.cpp

```diff
@@ -0,0 +1,645 @@
+bool scurves::move_from_pva_straight(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_straight(_t, time_scale, pos, vel, accel);
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_to_pva_straight(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_straight(_t, time_scale, pos, vel, accel);
+    pos -= _track;
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_time_pva_straight(float time, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    float scurve_P1, scurve_V1, scurve_A1, scurve_J1;
+    bool finish = update(time, scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_1 * scurve_P1;
+    vel += _delta_unit_1 * scurve_V1 * MIN(time_scale * 1.1f, 1.0f);
+    accel += _delta_unit_1 * scurve_A1 * time_scale;
+    return finish;
+}
+
+// Spline implementations
+
+// increment time pointer and return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_pva_spline(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_spline(_t, time_scale, pos, vel, accel);
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the destination
+bool scurves::move_to_pva_spline(float dt, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    advance_time(time_scale * dt);
+    bool finish = move_from_time_pva_spline(_t, time_scale, pos, vel, accel);
+    pos -= _track;
+    return finish;
+}
+
+// return the position, velocity and acceleration vectors relative to the origin
+bool scurves::move_from_time_pva_spline(float time, float time_scale, Vector3f &pos, Vector3f &vel, Vector3f &accel)
+{
+    bool finish;
+    float scurve_P1, scurve_V1, scurve_A1, scurve_J1;
+    float time_start = segment[7].start_time;
+    float time_mid = segment[num_segs - 1].start_time - segment[11].start_time;
+
+    update(MIN(time, segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_1 * scurve_P1;
+    vel += _delta_unit_1 * scurve_V1;
+    accel += _delta_unit_1 * scurve_A1;
+
+    update(MAX(time - time_start + segment[11].start_time, segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
+    pos += _delta_unit_2 * (scurve_P1 - segment[11].start_pos);
+    vel += _delta_unit_2 * scurve_V1;
+    accel += _delta_unit_2 * scurve_A1;
+
+    update(MIN(segment[11].start_time - (time - time_start * 2 - time_mid + segment[11].start_time), segment[11].start_time), scurve_J1, scurve_A1, scurve_V1, scurve_P1);
```

Comment:
It would be nice to get some defines instead of using constantss.  Or perhaps even better just explain what segment[11] is just here.

Reply:
There is a list of definitions at the top of the SCurve.cpp file now.
#define SEG_INIT        0
#define SEG_ACCEL_MAX   4
#define SEG_ACCEL_END   7
#define SEG_CHANGE_END  14
#define SEG_CONST       15
#define SEG_DECEL_END   22